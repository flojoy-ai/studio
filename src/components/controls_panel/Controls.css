.App-controls-panel {
    height: 100vh; 
}

.App-controls-header {
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  justify-content: flex-start;
  align-items: auto;
  align-content: flex-start; 
}

.ctrl-output {
  flex: 0 0 auto;
}

.ctrl-output, .ctrl-input {
  padding: 0;
  border: 0.5px solid #ccc;
  border-radius: 3px;
  position: relative;
  resize: both;
  cursor: pointer;
  /* -webkit-user-drag: element; */
  /* margin-top: 10px; */
}

.ctrl-output {
  margin: 20px 0 0 20px;
  min-height: 300px;
  height: 25%;
  width: 45%;
}

.ctrl-input {
  min-height: 150px;
  width: 100%;
 
}
.ctrl-header{
  display: flex;
  justify-content: flex-end;
  align-items: center;
  gap: 5px;
}
.ctrl-input-group {
  display:flex;
  gap:5px;
  flex-direction:column;
  padding-left:5px;
  padding-right:5px;
}
.ctrl-edit-btn {
  cursor: pointer;
}
.ctrl-close-btn {
  margin: 5px;
  cursor: pointer;
  position: absolute;
  top: 0;
  right: 0;
  border: 0;
  background: transparent;
  z-index: 99;
}

.ctrl-input-body {
  padding: 20px;
  display: flex;
  justify-content: center;
  align-items:center;
  height:90%;
  flex-wrap: wrap;
}

.ctrl-param {
  margin: 0;
  font-size: 15px;
}

.ctrl-close-btn:hover{
  color: red !important;
}

.ctrl-meta {
  font-size: 10px;
  text-align: left;
  position: absolute;
  bottom: 10px;
}

.input-header, .output-header {
  padding: 10px;
  display: inline-block;
}

.ctrl-inputs-sidebar, .input-header {
  min-width: 200px;
  max-width: 400px;
  width: 25%;  border-right: 1px solid #ccc;
  display: inline-block;
}

.ctrl-inputs-sidebar {
  height: 100%;
  padding: 10px;
}

.ctrl-outputs-canvas {
  display: flex;
  background-size: .75em .75em;  
  background-position: 0 -0.5em;
  flex-wrap: wrap;
  flex-direction: row;
  justify-content: flex-start;
  align-items: auto;
  align-content: flex-start;
  min-height: 800px;
}

.ctrl-outputs-container, .output-header {
  display: inline-block;
  width: 75%;
}
.ctrl-outputs-container {
  vertical-align: top;
  height: 100%;
}

.ctrl-picker-container {
  margin-top: 20px;
}

.ctrl-numeric-input {
  padding: 10px;
  width: 90%;
  margin: 0 10px;
  background-color: transparent;
  outline: none;
  border:'1px solid'
  ; border-radius: 10px;
}

.select-node {
  font-size: 12px;
  position: absolute;
  top: 0;
  left: 0;
  border: 0;
  width: 100%;
  border-bottom: 0.5px solid;
  margin-bottom: 10px;
  padding-right: 20px;
}

.ctrl-input label {
  margin-left: 10px;

}