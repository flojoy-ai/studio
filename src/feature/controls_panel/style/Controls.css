.App-controls-panel {
    height: 100vh; 
}

.App-controls-header {
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  justify-content: flex-start;
  align-items: auto;
  align-content: flex-start; 
}

.ctrl-output {
  flex: 0 0 auto;
}

.ctrl-output, .ctrl-input {
  padding: 0;
  border: 0.5px solid #ccc;
  border-radius: 3px;
  position: relative;
  resize: both;
  cursor: pointer;
  /* -webkit-user-drag: element; */
  /* margin-top: 10px; */
}

.ctrl-output {
  margin: 20px 0 0 20px;
  min-height: 300px;
  height: 25%;
  width: 45%;
}

.ctrl-input {
  min-height: 150px;
  width: 100%;
 
}
.ctrl-header{
  display: flex;
  justify-content: flex-end;
  align-items: center;
  gap: 5px;
}
.ctrl-input-group {
  display:flex;
  gap:5px;
  flex-direction:column;
  padding-left:5px;
  padding-right:5px;
}
.ctrl-edit-btn {
  cursor: pointer;
}
.ctrl-close-btn {
  margin: 5px;
  cursor: pointer;
  position: absolute;
  top: 0;
  right: 0;
  border: 0;
  background: transparent;
  z-index: 99;
}

.ctrl-input-body {
  padding: 20px;
  display: flex;
  justify-content: center;
  align-items:center;
  height:90%;
  flex-wrap: wrap;
  width: 100%;
}

.ctrl-input-body-file {
  /* padding: 20px; */
  display: grid;
  justify-content: center;
  align-items: center;
  height:90%;
  flex-wrap: wrap;
  width: 100%;
}


.ctrl-param {
  margin: 0;
  font-size: 15px;
}

.ctrl-close-btn:hover{
  color: red !important;
}

.ctrl-meta {
  font-size: 10px;
  text-align: left;
  position: absolute;
  bottom: 10px;
  left: 15px;
}

.input-header, .output-header {
  padding: 10px;
  display: inline-block;
}

.ctrl-inputs-sidebar, .input-header {
  min-width: 200px;
  max-width: 400px;
  width: 25%;  border-right: 1px solid #ccc;
  display: inline-block;
}

.ctrl-inputs-sidebar {
  height: 100%;
  padding: 10px;
}

.ctrl-outputs-canvas {
  display: flex;
  background-size: .75em .75em;  
  background-position: 0 -0.5em;
  flex-wrap: wrap;
  flex-direction: row;
  justify-content: flex-start;
  align-items: auto;
  align-content: flex-start;
  min-height: 800px;
}

.ctrl-outputs-container, .output-header {
  display: inline-block;
  width: 75%;
}
.ctrl-outputs-container {
  vertical-align: top;
  height: 100%;
}

.ctrl-picker-container {
  margin-top: 20px;
}

.ctrl-numeric-input {
  padding: 10px;
  width: 90%;
  margin: 0 10px;
  background-color: transparent;
  outline: none;
  border:'1px solid'
  ; border-radius: 10px;
}

.select-node {
  font-size: 12px;
  top: 0;
  left: 0;
  border: 0;
  width: 100%;
  border-bottom: 0.5px solid;
  margin-bottom: 10px;
  padding-right: 20px;
}

.select-plot-type {
  font-size: 12px;
  position: absolute;
  top: 0;
  left: 0;
  border: 0;
  width: 100%;
  border-bottom: 0.5px solid;
  margin-bottom: 10px;
  padding-right: 10px;
}

.ctrl-input label {
  margin-left: 10px;

}
.seven_segment_container {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;
  width: 100%;
  padding-bottom: 10px;
  min-width: 480;
  transform:scale(0.6)
}

 /* Extra small devices (phones, 600px and down) */
 @media only screen and (max-width: 600px) {
  .seven_segment_container{
    transform: scale(0.5);
  }
 }

 /* Small devices (portrait tablets and large phones, 600px and up) */
 @media only screen and (min-width: 600px) {
  .seven_segment_container{
    transform: scale(0.6);
  }
 }
 
 /* Medium devices (landscape tablets, 768px and up) */
 @media only screen and (min-width: 768px) {
  .seven_segment_container{
    transform: scale(0.5);
  }
 }
 
 /* Large devices (laptops/desktops, 992px and up) */
 /* @media only screen and (min-width: 992px) {...} */
 
 /* Extra large devices (large laptops and desktops, 1200px and up) */
 @media only screen and (min-width: 1200px) {
  .seven_segment_container{
    transform: scale(0.6);
  }
 } 