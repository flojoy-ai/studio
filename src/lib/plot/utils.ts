export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
