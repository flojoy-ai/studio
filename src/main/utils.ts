import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
