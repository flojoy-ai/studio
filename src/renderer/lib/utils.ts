import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";
import { z } from "zod";
import * as E from "fp-ts/Either";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tryParse<Z extends z.ZodTypeAny>(z: Z) {
  return (val: unknown) => {
    const res = z.safeParse(val);
    if (res.success) {
      return E.right(res.data as z.infer<Z>);
    }
    return E.left(res.error);
  };
}
