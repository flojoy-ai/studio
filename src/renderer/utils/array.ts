export function filterMap<T, U>(
  arr: Array<T>,
  mapPred: (x: T) => U | undefined | null,
) {
  return arr.reduce((acc: Array<U>, x: T) => {
    const val = mapPred(x);
    if (val !== undefined && val !== null) {
      acc.push(val);
    }

    return acc;
  }, []);
}
