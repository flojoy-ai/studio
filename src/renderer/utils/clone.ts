import { DeepMutable } from "@/renderer/types/util";

export const deepMutableClone = <T>(obj: T): DeepMutable<T> => {
  return JSON.parse(JSON.stringify(obj)) as DeepMutable<T>;
};
