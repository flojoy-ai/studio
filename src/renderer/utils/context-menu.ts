export const calculateContextMenuOffset = (
  clickX: number,
  clickY: number,
  flow: HTMLDivElement,
) => {
  // Calculate position of the context menu. We want to make sure it
  // doesn't get positioned off-screen.
  const pane = flow.getBoundingClientRect();

  const contextMenuHeight = 200;
  const paneToBlock = clickY - contextMenuHeight;

  let top: number | undefined = undefined;
  let bottom: number | undefined = undefined;

  if (paneToBlock < contextMenuHeight / 2) {
    top = paneToBlock;
  } else if (paneToBlock < contextMenuHeight) {
    if (pane.height - paneToBlock < contextMenuHeight) {
      top = contextMenuHeight - paneToBlock;
    } else {
      top = paneToBlock;
    }
  } else if (pane.height - paneToBlock < contextMenuHeight) {
    top = undefined;
    bottom = pane.height - paneToBlock;
  } else {
    top = paneToBlock;
  }

  return {
    top,
    left: clickX < pane.width - 200 ? clickX : undefined,
    right: clickX >= pane.width - 200 ? pane.width - clickX : undefined,
    bottom,
  };
};
