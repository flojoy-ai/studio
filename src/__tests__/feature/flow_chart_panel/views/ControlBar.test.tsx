import { screen } from "@testing-library/react";
import Controls from "@src/feature/flow_chart_panel/views/ControlBar";
import { renderWithTheme } from "@src/__tests__/__utils__/utils";

// Mock useFlowChartState hook
jest.mock("@src/hooks/useFlowChartState");
// Mock useSocket hook
jest.mock("@src/hooks/useSocket");
jest.mock("react-switch", () => ({
  __esModule: true,
  default: jest.fn(() => {
    return <div data-testid="react-switch" />;
  }),
}));
jest.mock("react-modal");
// Mock saveFlowChartToLocalStorage, saveAndRunFlowChartInServer, and cancelFlowChartRun functions
jest.mock("@src/services/FlowChartServices", () => ({
  saveFlowChartToLocalStorage: jest.fn(),
  saveAndRunFlowChartInServer: jest.fn(),
  cancelFlowChartRun: jest.fn(),
}));

// Mock KeyboardShortcutModal component
jest.mock("@src/feature/flow_chart_panel/views/KeyboardShortcutModal", () => ({
  __esModule: true,
  default: jest.fn(() => <div>KeyboardShortcutModal</div>),
}));
// Mock DropDown component
jest.mock("@src/feature/common/DropDown", () => ({
  __esModule: true,
  default: jest.fn(() => <div>Dropdown</div>),
}));

describe("Controls", () => {
  it("should render correctly", () => {
    const { container } = renderWithTheme(
      <Controls activeTab="visual" setOpenCtrlModal={jest.fn()} />
    );

    expect(screen.getByTestId("btn-play")).toBeInTheDocument();
    expect(screen.getByTestId("btn-play")).toBeDisabled();
    expect(screen.queryByTitle("Cancel Run")).not.toBeInTheDocument();
    expect(screen.queryByTestId("add-ctrl")).not.toBeInTheDocument();
    expect(screen.getByText("Dropdown")).toBeInTheDocument();
    expect(container).toMatchSnapshot("__main__");
  });
  it("should not show the Add node or add ctrl button when active tab is 'debug'", () => {
    renderWithTheme(
      <Controls activeTab="debug" setOpenCtrlModal={jest.fn()} />
    );
    expect(screen.queryByTestId("add-ctrl")).not.toBeInTheDocument();
  });
});
