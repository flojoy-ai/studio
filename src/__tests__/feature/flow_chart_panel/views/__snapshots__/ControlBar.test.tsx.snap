// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls should render correctly: __main__ 1`] = `
<div>
  <div
    class="mantine-1mo5lls"
  >
    <button
      class="mantine-8mgm8e"
      data-cy="btn-play"
      data-testid="btn-play"
      disabled=""
      title="Server is offline"
    >
      <svg
        fill="none"
        height="11"
        viewBox="0 0 9 11"
        width="9"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.5 4.63397C9.16667 5.01887 9.16667 5.98113 8.5 6.36603L1.75 10.2631C1.08333 10.648 0.25 10.1669 0.25 9.39711L0.25 1.60289C0.25 0.833085 1.08333 0.35196 1.75 0.73686L8.5 4.63397Z"
          fill="#99f5ff"
        />
      </svg>
      <span>
        Play
      </span>
    </button>
    <div>
      Dropdown
    </div>
    <div>
      KeyboardShortcutModal
    </div>
  </div>
  ,
</div>
`;
